File: app/src/main/java/br/fmu/bioreino/util/SessaoCategoriasUtil.java
package br.fmu.bioreino.util;

import android.content.res.Configuration;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.widget.AdapterView;
import android.widget.TextView;

import androidx.core.content.res.ResourcesCompat;
import androidx.recyclerview.widget.RecyclerView;

import br.fmu.bioreino.R;
import br.fmu.bioreino.adapter.ListaCategoriasAdapter;
import br.fmu.bioreino.adapter.SpinnerCategoriasAdapter;
import br.fmu.bioreino.componentes.SessaoCategoriaSpinner;

public class SessaoCategoriasUtil {

    private SessaoCategoriaSpinner spinner;
    private ListaCategoriasAdapter listaCategoriasAdapter;

    private final Drawable abertoLight;
    private final Drawable fechadoLight;
    private final Drawable abertoDark;
    private final Drawable fechadoDark;

    public Drawable backgroundSpinnerAberto;
    public Drawable backgroundSpinnerFechado;

    final SessaoCategoriaSpinner.SpinnerListener spinnerListener;
    final View view;

    public SessaoCategoriasUtil(SessaoCategoriaSpinner.SpinnerListener spinnerListener, View view) {
        this.spinnerListener = spinnerListener;
        this.view = view;

        this.abertoLight = ResourcesCompat.getDrawable(view.getResources(), R.drawable.bg_spinner_aberto_light, view.getContext().getTheme());
        this.fechadoLight = ResourcesCompat.getDrawable(view.getResources(), R.drawable.bg_spinner_fechado_light, view.getContext().getTheme());
        this.abertoDark = ResourcesCompat.getDrawable(view.getResources(), R.drawable.bg_spinner_aberto_dark, view.getContext().getTheme());
        this.fechadoDark = ResourcesCompat.getDrawable(view.getResources(), R.drawable.bg_spinner_fechado_dark, view.getContext().getTheme());
        atualizaTemaSeletor();

    }


    // Lista
    public void configuraSessaoCategorias(CategoriaInterface categoriaInterface) {
        configuraListaCategorias(categoriaInterface);
        configuraSpinner();
    }

    public void configuraListaCategorias(CategoriaInterface categoriaInterface) {
        RecyclerView categoriasRecyclerView = view.findViewById(R.id.sessao_categorias_lista);
        listaCategoriasAdapter = new ListaCategoriasAdapter(categoriaInterface);
        categoriasRecyclerView.setAdapter(listaCategoriasAdapter);
    }


    // Spinner
    private void configuraSpinner() {
        spinner = view.findViewById(R.id.sessao_categorias_spinner);
        SpinnerCategoriasAdapter spinnerAdapter = new SpinnerCategoriasAdapter(
                view.getResources().getStringArray(R.array.categorias)
        );

        spinner.setAdapter(spinnerAdapter);
        spinner.setSpinnerListener(spinnerListener);
        spinner.setBackground(backgroundSpinnerFechado);
        spinner.setSelection(0);
        setFiltroCategoria();
        listaCategoriasAdapter.getFilter().filter("Kids");
    }

    public void atualizaTemaSeletor() {
        int nightModeFlags = view
                .getContext().getApplicationContext()
                .getResources().getConfiguration().uiMode & Configuration.UI_MODE_NIGHT_MASK;

        switch (nightModeFlags) {
            case Configuration.UI_MODE_NIGHT_YES:
                backgroundSpinnerAberto = abertoDark;
                backgroundSpinnerFechado = fechadoDark;
                break;

            case Configuration.UI_MODE_NIGHT_NO:
            case Configuration.UI_MODE_NIGHT_UNDEFINED:
                backgroundSpinnerAberto = abertoLight;
                backgroundSpinnerFechado = fechadoLight;
                break;
        }

    }

    public void setFiltroCategoria() {
        spinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
            @Override
            public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
                filtraLista();
            }

            @Override
            public void onNothingSelected(AdapterView<?> adapterView) {

            }
        });
    }

    private void filtraLista() {
        View opcaoSelecionada = spinner.getSelectedView();
        if (opcaoSelecionada != null) {
            TextView texto = opcaoSelecionada.findViewById(R.id.item_plano_categoria_label);
            listaCategoriasAdapter.getFilter().filter(texto.getText());
        }
    }

}

File: app/src/main/java/br/fmu/bioreino/viewModel/util/SessaoCategoriasUtil.kt
package br.fmu.bioreino.viewModel.util

import android.content.res.Configuration
import android.graphics.drawable.Drawable
import android.view.View
import android.widget.AdapterView
import android.widget.TextView
import androidx.core.content.res.ResourcesCompat
import androidx.recyclerview.widget.RecyclerView
import br.fmu.bioreino.R
import br.fmu.bioreino.viewModel.adapter.ListaCategoriasAdapter
import br.fmu.bioreino.viewModel.adapter.SpinnerCategoriasAdapter
import br.fmu.bioreino.model.componentes.CategoriesSessionSpinner
import br.fmu.bioreino.model.componentes.CategoriesSessionSpinner.CategoriesSpinnerListener

class SessaoCategoriasUtil(private val categoriesSpinnerListener: CategoriesSpinnerListener, val view: View) {
    private var listaCategoriasAdapter: ListaCategoriasAdapter? = null

    private val abertoLight: Drawable? = ResourcesCompat.getDrawable(
        view.resources,
        R.drawable.bg_spinner_aberto_light,
        view.context.theme
    )
    private val fechadoLight: Drawable? = ResourcesCompat.getDrawable(
        view.resources,
        R.drawable.bg_spinner_fechado_light,
        view.context.theme
    )
    private val abertoDark: Drawable? = ResourcesCompat.getDrawable(
        view.resources,
        R.drawable.bg_spinner_aberto_dark,
        view.context.theme
    )
    private val fechadoDark: Drawable? = ResourcesCompat.getDrawable(
        view.resources,
        R.drawable.bg_spinner_fechado_dark,
        view.context.theme
    )

    @JvmField
    var backgroundSpinnerAberto: Drawable? = null

    @JvmField
    var backgroundSpinnerFechado: Drawable? = null

    init {
        atualizaTemaSeletor()
    }

    // Lista
    fun configuraSessaoCategorias(categoriaInterface: CategoriaInterface) {
        configuraListaCategorias(categoriaInterface)
        configuraSpinner()
    }

    private fun configuraListaCategorias(categoriaInterface: CategoriaInterface) {
        val categoriasRecyclerView = view.findViewById<RecyclerView>(R.id.sessao_categorias_lista)
        listaCategoriasAdapter = ListaCategoriasAdapter(categoriaInterface)
        categoriasRecyclerView.adapter = listaCategoriasAdapter
    }

    // Spinner
    private fun configuraSpinner() {
        val spinner: CategoriesSessionSpinner = view.findViewById(R.id.sessao_categorias_spinner)
        val spinnerAdapter = SpinnerCategoriasAdapter(
            view.resources.getStringArray(R.array.categorias)
        )
        spinner.adapter = spinnerAdapter
        spinner.categoriesSpinnerListener = categoriesSpinnerListener
        spinner.background = backgroundSpinnerFechado
        spinner.setSelection(0)
        setFiltroCategoria(spinner)
        listaCategoriasAdapter!!.filter.filter("Kids")
    }

    fun atualizaTemaSeletor() {
        val nightModeFlags = view
            .context.applicationContext
            .resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
        when (nightModeFlags) {
            Configuration.UI_MODE_NIGHT_YES -> {
                backgroundSpinnerAberto = abertoDark
                backgroundSpinnerFechado = fechadoDark
            }
            Configuration.UI_MODE_NIGHT_NO, Configuration.UI_MODE_NIGHT_UNDEFINED -> {
                backgroundSpinnerAberto = abertoLight
                backgroundSpinnerFechado = fechadoLight
            }
        }
    }

    private fun setFiltroCategoria(spinner: CategoriesSessionSpinner) {
        spinner.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
            override fun onItemSelected(adapterView: AdapterView<*>?, view: View?, i: Int, l: Long) {
                filtraLista(spinner)
            }

            override fun onNothingSelected(adapterView: AdapterView<*>?) {}
        }
    }

    private fun filtraLista(spinner: CategoriesSessionSpinner) {
        val opcaoSelecionada = spinner.selectedView
        if (opcaoSelecionada != null) {
            val texto = opcaoSelecionada.findViewById<TextView>(R.id.item_plano_categoria_label)
            listaCategoriasAdapter!!.filter.filter(texto.text)
        }
    }
}